### PR TITLE
#3745 - Modify process that reads SIN & CRA verification response files

### PR DESCRIPTION
**Pull request by bidyashish**
***Acceptance Criteria***
- [x] Jobs must continue processing remaining records in a file after it fails to process a record. 
- [x] SIN and CRA response files will need to be archived after processing even where there are failed records. (IMB staff will need to monitor the jobs to confirm if errors occur processing SIMS response files and address them accordingly)
- [x] Disable error reporting when SIMS Reference_IDX fails to match.
- [x] If records are identified as SIMS records (based on the project area information being present), errors will be generated as usual.
  - [x] If records are identified as SFAS records (based on the project area information NOT being present), records should be skipped.
- [x] For the SIN validation ([source reference](https://github.com/bcgov/SIMS/blob/8864eaa8a5f62360f5e3c91c6de06f319e008250/sources/packages/backend/libs/integrations/src/services/sin-validation/sin-validation.service.ts#L119)), if the `referenceIndex` is not found on SIMS, assume the record is from SFAS and do not log as an error.
- [x] For income verifications, the records should already be skipped. Execute a manual test to ensure that a file with records that do not contain the `VERIFICATION_ID` in the free project area is skipped. Consider creating an E2E to execute the test.
- [x] Demo for both cases with a sample file.

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    db = createE2EDataSources(dataSource);
+    sftpClientMock = sshClientMock;
+    processor = app.get(CRAResponseIntegrationScheduler);
+  });
+
+  beforeEach(async () => {
+    jest.clearAllMocks();
+  });
+
+  it.only("should process CRA response file", async () => {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:55]: Please remove the `only`.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+} from "@sims/test-utils/mocks";
+import { Job } from "bull";
+import { ApplicationStatus } from "@sims/sims-db";
+
+const CRA_FILENAME = "CRA_200_PBCSA00000.TXT";
+
+const padWithLeadingZeros = (num: number): string => {
+  // Pad the number with leading zeros to make it 9 digits long
+  return num.toString().padStart(9, "0");
+};
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:31]: Helper methods are usually created at the end of the tests, inside the `describe` they belong.
For this one, I would refrain from creating the method, but keeping it will no be a blocker if moved to the right place and added method comments.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:1]: All comments should have periods.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      { initialValues: { dateReceived: null } },
+    );
+    await db.craIncomeVerification.save([studentCRAIncomeVerification]);
+    // Queued job.
+    const job = createMock<Job<void>>();
+    mockDownloadFiles(sftpClientMock, [CRA_FILENAME]);
+
+    mockDownloadFiles(sftpClientMock, [CRA_FILENAME], (fileContent: string) => {
+      const file = getStructuredRecords(fileContent);
+      const line2 = file.records[2]; // Get the 4th item (index 3) in the array
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:79]: The `file.records` contains only records (no header) and the comment is misleading because it is actually getting 3rd item in the array, index 2. Does it make sense?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+      // Split the record on colon and change 9 digits after colon
+      const [beforeColon, afterColon] = line2.split(":");
+      const newVerificationId = padWithLeadingZeros(
+        studentCRAIncomeVerification.id,
+      );
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:85]: Can we use one replacement strategy similar to the one below?
https://github.com/bcgov/SIMS/blob/4e3b55ab77ffb49cf25f35eaa0fd0f8650e38e3f/sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ece-response/_tests_/ece-response-integration.scheduler.e2e-spec.ts#L136-L145

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-receive-files/CRA_200_PBCSA00000.TXT
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-receive-files/CRA_200_PBCSA00000.TXT:1]: Is it, not the idea that this file would contain the result for 2 income requests where one would be ignored to simulate the SFAS one?
For instance, lines 4 and 5 below represent a `ResponseStatusRecord` and a `TotalIncome` from SIMS. Should we have the same for SFAS?
image

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      db = createE2EDataSources(dataSource);
+      sftpClientMock = sshClientMock;
+      processor = app.get(SINValidationResponseIntegrationScheduler);
+    });
+
+    beforeEach(async () => {
+      jest.clearAllMocks();
+    });
+
+    it.only("should process SIN validation response file", async () => {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts:54]: Please remove the ".only".

File: sources/packages/backend/libs/integrations/src/services/sin-validation/sin-validation.service.ts [hunk 6]
```diff
@@ -121,9 +121,10 @@ export class SINValidationService extends RecordDataModelService<SINValidation>
           .getOne();
 
         if (!existingValidation) {
-          throw new Error(
-            `Not able to find the SIN validation id ${validationResponse.referenceIndex} to be updated with the ESDC response.`,
-          );
+          return {
+            operationDescription: `Not able to find the SIN validation on line number ${validationResponse.lineNumber} to be updated with the ESDC response.`,
+            record: null,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/sin-validation/sin-validation.service.ts:126]: The record should be declared optional.
```
export interface SINValidationUpdateResult {
  record?: SINValidation;
  operationDescription: string;
}
```
And this can be returned as
```ts
return {
   operationDescription: `...`,
};
```

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        SIN_VALIDATION_FILENAME,
+      );
+
+      // Assert
+      expect(processResult).toStrictEqual([
+        {
+          processSummary: [
+            `Processing file ${downloadedFile}.`,
+            "File contains 2 SIN validations.",
+            "Processed SIN validation record from line 2: No SIN validation was updated because the record id is already present and this is not the most updated.",
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts:99]: Please ensure the data setup will lead to the below message. It is going to a "non-happy path" probably because `dateReceived` is set.

This one the the happy, path.
```
SIN validation record updated.
```

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts:1]: Please ensure that any applicable comment in the other E2E will be applied here also if required.

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work with the E2Es. Please take a look at the comments.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    db = createE2EDataSources(dataSource);
+    sftpClientMock = sshClientMock;
+    processor = app.get(CRAResponseIntegrationScheduler);
+  });
+
+  beforeEach(async () => {
+    jest.clearAllMocks();
+  });
+
+  it("Should process CRA response file to update the income verification record.", async () => {
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:55]: Please use "Should/when" for the test descriptions.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+      { student },
+      { applicationStatus: ApplicationStatus.InProgress },
+    );
+
+    // Create CRA income verifications for student.
+    const studentCRAIncomeVerification = createFakeCRAIncomeVerification(
+      {
+        application,
+      },
+      { initialValues: { dateReceived: null } },
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:70]: if you don't pass `dateReceived` it'll be null. 
image

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    );
+    await db.craIncomeVerification.save([studentCRAIncomeVerification]);
+    // Queued job.
+    const job = createMock<Job<void>>();
+    mockDownloadFiles(sftpClientMock, [CRA_FILENAME]);
+
+    mockDownloadFiles(sftpClientMock, [CRA_FILENAME], (fileContent: string) => {
+      const file = getStructuredRecords(fileContent);
+      const line2 = file.records[2]; // Get the 4th item (index 3) in the array
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:79]: Please update the comment.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+      // Split the record on colon and change 9 digits after colon
+      const [beforeColon, afterColon] = line2.split(":");
+      const newVerificationId = padWithLeadingZeros(
+        studentCRAIncomeVerification.id,
+      );
+      const updatedRecord4 = `${beforeColon}:${newVerificationId}${afterColon.substring(
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:86]: Update the variable name or change it to a generic one as it's important for the test to know if third or fourth record.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+            "Processed SIN validation record from line 3: Not able to find the SIN validation on line number 3 to be updated with the ESDC response.",
+          ],
+          errorsSummary: [],
+        },
+      ]);
+    });
+
+    it("Should process SIN validation response file to update the SIN validation record.", async () => {
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts:107]: Same as https://github.com/bcgov/SIMS/pull/3835/files#r1821245787

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:31]: The PR does not seem to be updated with the change.

image

Please read the comment given carefully to ensure that nothing is missing.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts (hunk 8 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts]: By "Should/when" I meant something like "**Should** update the income verification records **when** processing CRA response file."

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      { student },
+      { applicationStatus: ApplicationStatus.InProgress },
+    );
+
+    // Create CRA income verifications for student.
+    const studentCRAIncomeVerification = createFakeCRAIncomeVerification(
+      {
+        application,
+      },
+      { initialValues: {} },
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:65]: You don't need to pass the options if it's empty.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:80]: To be clear, the comment was not targeting the removal of the ":" but the use of the replace to adjust the file content.

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cra-integration/_tests_/cra-response-integration.scheduler.e2e-spec.ts:80]: The below is the expected code. Tested locally and it works.
```ts
mockDownloadFiles(sftpClientMock, [CRA_FILENAME], (fileContent: string) => {
      const file = getStructuredRecords(fileContent);
      file.records[2] = file.records[2].replace(
        "CRA_INCOME_VERIFICATION",
        studentCRAIncomeVerification.id.toString().padStart(9, "0"),
      );
      return createFileFromStructuredRecords(file);
    });
```

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+            "File contains 2 SIN validations.",
+            "Processed SIN validation record from line 2: No SIN validation was updated because the record id is already present and this is not the most updated.",
+            "Processed SIN validation record from line 3: Not able to find the SIN validation on line number 3 to be updated with the ESDC response.",
+          ],
+          errorsSummary: [],
+        },
+      ]);
+    });
+
+    it("Should update the SIN validation record and skip the update when the dateReceived is null.", async () => {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts:102]: Pleas update to Should/When as below.
```
Should update one SIN validation record and skip one when one SIN response is from SIMS and the other is from SFAS.
```

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+          const [record1] = file.records;
+
+          // Update the first record with validSinStudent's padded ID
+          const paddedId1 = padWithLeadingZeros(
+            validSinStudent.sinValidation.id,
+          );
+          file.records[0] =
+            record1.substring(0, 3) + paddedId1 + record1.substring(12);
+          return createFileFromStructuredRecords(file);
+        },
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/sin-validation-integration/_tests_/sin-validation-process-response-integration.scheduler.e2e-spec.ts:77]: Yeah, got it. Please follow the same from here: https://github.com/bcgov/SIMS/pull/3835#discussion_r1821629862

File: sources/packages/backend/libs/sims-db/src/entities/sin-validation.model.ts [hunk 16]
```diff
@@ -47,7 +47,7 @@ export class SINValidation extends RecordDataModel {
     type: "timestamptz",
     nullable: true,
   })
-  dateReceived?: Date;
+  dateReceived?: Date | null;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/sims-db/src/entities/sin-validation.model.ts:50]: The dateReceived is already marked as nullable and declared as "dateReceived?". Please revert this change.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, looks good 👍

**Review by andrepestana-aot — APPROVED** [review]
Looks good. Thanks for doing the changes!
